File: python_sdk/src/reality_capture/specifications/tile_map_optimization.py
from pydantic import BaseModel, Field, constr
from typing import Optional
from enum import Enum


class TileMapOptimizationInputs(BaseModel):
    tile_maps: list[str] = Field(description="Reality data Ids of tile maps to convert", alias="tileMaps")


class TileMapOptimizationFormat(Enum):
    XYZ_TILE_MAP = "XYZTileMap"


class TileMapImageFormat(Enum):
    JPG = "JPG"
    PNG = "PNG"


class TileMapOptimizationOptions(BaseModel):
    out_format: Optional[TileMapOptimizationFormat] = Field(None, description="Output format for the conversion.",
                                                            alias="format")
    input_crs: Optional[str] = Field(None, description="CRS for the input data", alias="inputCrs")
    output_crs: Optional[str] = Field(None, description="CRS for the output data", alias="outputCrs")
    top_level: Optional[int] = Field(None, description="Top level of the tile map to generate.", alias="topLevel")
    bottom_level: Optional[int] = Field(None, description="Bottom level of the tile map to generate.",
                                        alias="bottomLevel")
    image_format: Optional[TileMapImageFormat] = Field(None, alias="imageFormat", description="Image format")
    jpg_quality: Optional[int] = Field(None, description="Quality of JPG tiles to generate", alias="jpgQuality",
                                       ge=10, le=99)
    background_color: Optional[constr(pattern=r'^#[a-fA-F0-9]{6}$')] = (
        Field(None, description="Background color to use for tiles", alias="backgroundColor"))


class TileMapOptimizationSpecificationsCreate(BaseModel):
    inputs: TileMapOptimizationInputs = Field(description="Inputs")
    options: Optional[TileMapOptimizationOptions] = Field(None, description="Options")


class TileMapOptimizationSpecifications(BaseModel):
    inputs: TileMapOptimizationInputs = Field(description="Inputs")
    outputs: str = Field(description="Reality Data id of the tile map")
    options: Optional[TileMapOptimizationOptions] = Field(None, description="Options")

File: typescript/packages/reality-capture/src/specifications/tile_map_optimization.ts
import { z } from "zod";

export const TileMapOptimizationInputsSchema = z.object({
  tileMaps: z
    .array(z.string())
    .describe("Reality data Ids of tile maps to convert"),
});
export type TileMapOptimizationInputs = z.infer<
  typeof TileMapOptimizationInputsSchema
>;

export enum TileMapOptimizationFormat {
  XYZ_TILE_MAP = "XYZTileMap",
}

export enum TileMapImageFormat {
  JPG = "JPG",
  PNG = "PNG",
}

export const TileMapOptimizationOptionsSchema = z.object({
  outputFormat: z
    .nativeEnum(TileMapOptimizationFormat)
    .describe("Output format for the conversion.")
    .optional(),
  inputCrs: z.string().describe("CRS for the input data").optional(),
  outputCrs: z.string().describe("CRS for the output data").optional(),
  topLevel: z
    .number()
    .describe("Top level of the tile map to generate.")
    .optional(),
  bottomLevel: z
    .number()
    .describe("Bottom level of the tile map to generate.")
    .optional(),
  imageFormat: z
    .nativeEnum(TileMapImageFormat)
    .describe("Image format")
    .optional(),
  jpgQuality: z
    .number()
    .min(10)
    .max(99)
    .describe("Quality of JPG tiles to generate")
    .optional(),
  backgroundColor: z
    .string()
    .regex(/^#[a-fA-F0-9]{6}$/)
    .describe("Background color to use for tiles")
    .optional(),
});
export type TileMapOptimizationOptions = z.infer<
  typeof TileMapOptimizationOptionsSchema
>;

export const TileMapOptimizationSpecificationsCreateSchema = z.object({
  inputs: TileMapOptimizationInputsSchema.describe("Inputs"),
  options: TileMapOptimizationOptionsSchema.describe("Options").optional(),
});
export type TileMapOptimizationSpecificationsCreate = z.infer<
  typeof TileMapOptimizationSpecificationsCreateSchema
>;

export const TileMapOptimizationSpecificationsSchema = z.object({
  inputs: TileMapOptimizationInputsSchema.describe("Inputs"),
  output: z.string().describe("Reality Data id of the tile map"),
  options: TileMapOptimizationOptionsSchema.describe("Options").optional(),
});
export type TileMapOptimizationSpecifications = z.infer<
  typeof TileMapOptimizationSpecificationsSchema
>;
